(CODEMGMT-171) Use forge libraries for installing modules from the Forge

adrienthebo opened the pull request.

adrienthebo committed:
Add #versions and #latest_version to V3::Module

adrienthebo committed:
[forge/module_release] Update for current unpacker API

adrienthebo committed:
[module/forge] (maint) Don't use SemVer for module versions

The SemVer class was brought into r10k with the assumption that it would
be used for dependency resolution, but the Semantic library used for
version parsing has superseded the SemVer class. It provides no value
and complicates the code so this commit removes the conversions.

adrienthebo committed:
[module/forge] (maint) spec coverage for #expected_version

adrienthebo committed:
[module/forge] Use V3::Module for getting latest module version

adrienthebo committed:
[forge/module_release] add logging when performing actions

adrienthebo committed:
[forge/module_release] Add default temp paths

adrienthebo committed:
[module/forge] Use new puppet forge libraries for installation

adrienthebo committed:
[r10k/semver] Deprecate SemVer class prior to 2.0 removal

adreyer commented on lib/r10k/forge/module_release.rb:
```diff
@@ -29,8 +39,9 @@ def initialize(full_name, version)
-        #@download_path = Pathname.new(R10K::Settings[:forge][:module_cache]) + "#{@forge_release.slug}.tgz"
-        #@unpack_path   = Pathname.new(R10K::Settings[:forge][:unpack_tmpdir]) + @forge_release.slug
+
+        @download_path = Pathname.new(Dir.mktmpdir) + (slug + '.tar.gz')
```
Should this use a standard location?

mktmpdir can make debugging hard.

adrienthebo replied:
In the future, yes - rather than go down a rabbit hole of dealing with settings I've chosen to just use a tmpdir to get this merged and then revisit this in a future issue/pull request.

adrienthebo committed:
[puppetfile] Default mod args to nil

This only affects Forge modules as all other module types use a hash of
arguments. By changing the default to nil this allows Forge modules to
have easy short circuit logic for determining the appropriate module
version.

adrienthebo committed:
[module/forge] Remove Ruby 2.2 compat fix

Since SemVer objects are no longer used by this class, we can safely
compare values against :latest instead of doing a class type check
first.

andersonmills commented on lib/r10k/module/forge.rb:
```diff
@@ -143,28 +135,7 @@ def reinstall
-    repo = R10K::ModuleRepository::Forge.new
-    expected = repo.latest_version(title)
-    @expected_version = R10K::SemVer.new(expected)
-  end
```
I like that you're able to get rid of this method. :)

andersonmills commented on spec/unit/puppet_forge/v3/module_spec.rb:
```diff
@@ -3,6 +3,35 @@
+      }
+  end
+
+  let(:deleted_releases) do
```
This is nitpicky, but can we change this helper method name to :with_deleted_releases or :releases_including_deletes? The name caused me a bit of figuring out what the test was actually doing.

adrienthebo replied:
Fixed.
